PSRabbitMq/ConvertFrom-RabbitMqDelivery.ps1

```diff
@@ -11,26 +11,93 @@
     .PARAMETER Delivery
         RabbitMq Delivery to parse.
 
+    .PARAMETER IncludeEnvelope
+        Include the Message envelope (Metadata) of the message. If ommited, only 
+        the payload (body of the message) is returned
+
     .EXAMPLE
         ConvertFrom-RabbitMqDelivery -Delivery $Delivery
     #>
     param(
-        [RabbitMQ.Client.Events.BasicDeliverEventArgs]$Delivery
+        [RabbitMQ.Client.Events.BasicDeliverEventArgs]$Delivery,
+        [switch]$IncludeEnvelope
     )
     switch($Delivery.BasicProperties.ContentType) {
-        "text/plain" {
-            [Text.Encoding]::UTF8.GetString($Delivery.Body)
+        'text/plain' {
+            $Payload = [Text.Encoding]::UTF8.GetString($Delivery.Body)
         }
-        "application/clixml+xml" {
+        'application/clixml+xml' {
             $XmlBody = [Text.Encoding]::UTF8.GetString($Delivery.Body)
-            [System.Management.Automation.PSSerializer]::DeserializeAsList($XmlBody)
+            try
+            {
+                $deserialized = [System.Management.Automation.PSSerializer]::DeserializeAsList($XmlBody)
+            }
+            catch
+            {
+                #This is for V2 clients...
+                $TempFile = [io.path]::GetTempFileName()
+                try
+                {
+                    $null = New-Item -Name (Split-Path -Leaf $TempFile) -Value $XmlBody -ItemType File -Path (split-path $TempFile -Parent) -Force
+                    $deserialized = Import-Clixml -Path $TempFile
+                    $deserialized = [IO.File]::ReadAllLines($TempFile, [Text.Encoding]::UTF8)
+                }
+                finally
+                {
+                    if( (Test-Path -Path $TempFile) )
+                    {
+                        Remove-Item -Path $TempFile -Force
+                    }
+                }
+            }
+            $Payload = $deserialized
         }
-        "application/json" {
+        'application/json' {
             $JsonBody = [Text.Encoding]::UTF8.GetString($Delivery.Body)
-            ConvertFrom-Json $JsonBody
+            try
+            {
+                $Payload = ConvertFrom-Json $JsonBody
+            }
+            catch
+            {
+                Write-Error 'Invalid JSON. Returning String'
+                $Payload = $JsonBody
+            }
+        }
+        'text/xml' {
+            $Payload = [xml]([Text.Encoding]::UTF8.GetString($Delivery.Body))
         }
         default {
-            [Text.Encoding]::UTF8.GetString($Delivery.Body)
+            $Payload = [Text.Encoding]::UTF8.GetString($Delivery.Body)
+        }
+    }
+
+    if (-not $IncludeEnvelope)
+    {
+        return $Payload
+    }
+    else {
+        return [PSCustomObject][ordered]@{

```

Comment:
I'm not sure about the value of this envelope object, it looks like it's basically doing the same thing as:

`Add-Member -InputObject $Delivery -MemberType NoteProperty -Name Payload -Value $Payload`

Maybe instead of re-creating this object it would be better to add a -Raw switch and return the unmodified $Delivery object and leave it to the client to do the ConvertFrom-RabbitMqDelivery to get the payload?